### PR TITLE
MLPerf BERT: Main training loop

### PR DESCRIPTION
**Pull request by Eliulm**
https://wandb.ai/wahl/MLPerf/runs/nhhsj33t?nw=nwuserwahl

Command to run:
`DEBUG=2 DEFAULT_FLOAT=half GPUS=6 BEAM=4 CKPT=1 CKPT_DIR=~/ckpts-final TRAIN_STEPS=200000 OPT_BASE_LEARNING_RATE=0.0001 MODEL=bert PYTHONPATH=. BASEDIR=/raid/datasets/wiki python3 examples/mlperf/model_train.py`

### CODE REVIEW

File: examples/mlperf/dataloader.py [hunk 1]
```diff
@@ -155,22 +158,53 @@ def process_batch_bert(data: List[dict]) -> dict[str, Tensor]:
...
-      blob.extend(load_bert_file(files.pop(0)))
-      if not files: 
-        if val: files = get_val_files()
-        else: end = True # End of train data - avoid pop on empty file list
-    if len(blob) >= BS: # if last train step does not have enough for a full batch
-      yield process_batch_bert(blob[:BS])
-      blob = blob[BS:]
+def shuffle_parts(file_paths: List[str]) -> List[str]:
+  parts = list(set(map(lambda f: int(Path(f).name.split('_')[0]), file_paths))) # Unique part ids (e.g. 0-499)
+  random.Random().shuffle(parts)
```
Comment (chenyuxyz) [inline examples/mlperf/dataloader.py:163]: just random.shuffle should work too, do you need the instance?

File: examples/mlperf/dataloader.py [hunk 2]
```diff
@@ -155,22 +158,53 @@ def process_batch_bert(data: List[dict]) -> dict[str, Tensor]:
...
+  shuffled_parts = []
+  for p in parts:
+    shuffled_parts.extend([f for f in file_paths if int(Path(f).name.split("_")[0]) == p])
```
Comment (chenyuxyz) [inline examples/mlperf/dataloader.py:166]: i don't understand what this part if trying to do. is it filtering something but it's O(N^2) instead of O(N)?

File: examples/mlperf/dataloader.py [hunk 3]
```diff
@@ -155,22 +158,53 @@ def process_batch_bert(data: List[dict]) -> dict[str, Tensor]:
...
+  return shuffled_parts
+
+def random_sample(data: List[str]):
+  index = random.randint(0, len(data) - 1)
+  selected_sample = data[index]
+  return selected_sample, index
+
+def load_datasample(fn:Tuple[str, int]) -> List[dict]:
```
Comment (chenyuxyz) [inline examples/mlperf/dataloader.py:174]: prefer naming `fn[0]` and `fn[1]` explicitly

File: examples/mlperf/dataloader.py [hunk 4]
```diff
@@ -1,12 +1,14 @@
...
 import numpy as np
 from PIL import Image
 from tqdm import tqdm
-import pickle
 from tinygrad import dtypes, Tensor
 from tinygrad.helpers import getenv, prod, Timing, Context
-from multiprocessing import Queue, Process, shared_memory, connection, Lock, cpu_count
+from multiprocessing import Queue, Process, shared_memory, connection, Lock, cpu_count, Pool
+from functools import lru_cache
+from itertools import chain
```
Comment (chenyuxyz) [inline examples/mlperf/dataloader.py:11]: just `import functools, itertools` for consistency

File: examples/mlperf/helpers.py [hunk 5]
```diff
@@ -190,3 +192,89 @@ def get_bert_qa_prediction(features, example, start_end_logits):
...
+    config = json.load(f)
+  return BertForMLPerf(
+    config["hidden_size"],
+    config["intermediate_size"], 
+    config["max_position_embeddings"], 
+    config["num_attention_heads"], 
+    config["num_hidden_layers"], 
+    config["type_vocab_size"], 
+    config["vocab_size"],
+    # 0.0, 0.0,
```
Comment (chenyuxyz) [inline examples/mlperf/helpers.py:216]: remove this 0.0

File: examples/mlperf/model_train.py [hunk 6]
```diff
@@ -257,8 +256,248 @@ def train_rnnt():
...
+
+  GPUS = config["GPUS"] = [f"{Device.DEFAULT}:{i}" for i in range(getenv("GPUS", 1))]
+  print(f"Training on {GPUS}")
+  for x in GPUS: Device[x]
+  seed = config["seed"] = getenv("SEED", 12345)
+
+  # ** hyperparameters **
+  epochs             = config["epochs"]                 = getenv("EPOCHS", 1)
+  BS                 = config["GLOBAL_BATCH_SIZE"]      = getenv("BS", 4 * len(GPUS)) # FP32 6 GPUS -> BS24
+  EVAL_BS            = config["EVAL_BS"]                = getenv("EVAL_BS", 8)
```
Comment (chenyuxyz) [inline examples/mlperf/model_train.py:275]: 8 is not multiple of 6? or `2 * len(GPUS)`

**Review by chenyuxyz — COMMENTED** [review]
some minor comments, you can start the run after addressing these

File: examples/mlperf/dataloader.py (hunk 1 above)
Comment (Eliulm) [reply examples/mlperf/dataloader.py:163]: `random.shuffle` will use the seed we set earlier. This has to be different each run.

File: examples/mlperf/dataloader.py (hunk 2 above)
Comment (Eliulm) [reply examples/mlperf/dataloader.py:166]: So mlperf shuffles all the parts first. Since the file_paths are pointing to individual files, we have to iterate over all files and check whether they belong to the current part. I will think about something that is not polynomial in runtime because I just noticed that doing that with all 500 parts takes forever.

File: examples/mlperf/dataloader.py (hunk 1 above)
Comment (chenyuxyz) [reply examples/mlperf/dataloader.py:167]: even if it uses the seed we set earlier, it still generates different shuffle right? (unless we reseed it with the same seed)

it's always preferred to make these explicit and not implicitly recreate the class to hide the issue

Comment (Eliulm) [reply examples/mlperf/dataloader.py:167]: Well the shuffle only ever happens once at the start so from my understanding, each run would have the same seed. We only ever do one epoch for training

Comment (chenyuxyz) [reply examples/mlperf/dataloader.py:167]: we will set different seed for each run, the point is that that seed should be the only random seed

File: examples/mlperf/model_train.py [hunk 7]
```diff
@@ -257,8 +256,248 @@ def train_rnnt():
...
+    load_training_state(model, optimizer_group, scheduler, safe_load(ckpt))
+    start_step = scheduler.epoch_counter.numpy().item()
+    print(f"resuming from {ckpt} at step {start_step}")
+
+  # ** init wandb **
+  WANDB = getenv("WANDB")
+  if WANDB:
+    import wandb
+    wandb_args = {"id": wandb_id, "resume": "must"} if (wandb_id := getenv("WANDB_RESUME", "")) else {}
+    wandb.init(config=config, **wandb_args, project="MLPerf")
```
Comment (chenyuxyz) [inline examples/mlperf/model_train.py:343]: add bert to project name

**Comment by chenyuxyz** [comment]
can you confirm if the training is typically done with dropout, and if the current implementation uses dropout (i don't think it is)?

File: examples/mlperf/initializers.py [hunk 8]
```diff
@@ -33,3 +34,35 @@ def __init__(self, in_features, out_features, bias=True):
...
+class LayerNormBert:
+  def __init__(self, normalized_shape:Union[int, Tuple[int, ...]], eps:float=1e-12, elementwise_affine:bool=True):
+    self.normalized_shape = (normalized_shape,) if isinstance(normalized_shape, int) else tuple(normalized_shape)
+    self.axis, self.eps, self.elementwise_affine = tuple(-1-i for i in range(len(self.normalized_shape))), eps, elementwise_affine
+    self.weight, self.bias = (Tensor.ones(*self.normalized_shape, dtype=dtypes.float32), Tensor.zeros(*self.normalized_shape, dtype=dtypes.float32)) if elementwise_affine else (None, None)
+
+  def __call__(self, x:Tensor):
+    assert self.normalized_shape == x.shape[-len(self.normalized_shape):], f"last dimensions of {x.shape} must match {self.normalized_shape}"
+    xn = x.cast(dtypes.float32).layernorm(eps=self.eps, axis=self.axis).cast(dtypes.default_float)
+    if not self.elementwise_affine: return xn.cast(x.dtype)
```
Comment (chenyuxyz) [inline examples/mlperf/initializers.py:67]: why is this line different? don't think it need cast now

File: examples/mlperf/model_train.py [hunk 9]
```diff
@@ -259,8 +258,250 @@ def train_rnnt():
...
+      wandb.log({"lr": optimizer.lr.numpy(), "train/loss": loss, "train/step_time": cl - st,
+                  "train/python_time": pt - st, "train/data_time": dt - pt, "train/cl_time": cl - dt,
+                  "train/GFLOPS": GlobalCounters.global_ops * 1e-9 / (cl - st)})
+
+    train_data, next_data = next_data, None
+    i += 1
+
+    if i == BENCHMARK:
+      median_step_time = sorted(step_times)[(BENCHMARK + 1) // 2]  # in seconds
+      estimated_total_hours = median_step_time * train_steps * epochs / 60 / 60
```
Comment (chenyuxyz) [inline examples/mlperf/model_train.py:432]: this has error and was fixed in resnet estimation

File: examples/mlperf/initializers.py (hunk 8 above)
Comment (Eliulm) [reply examples/mlperf/initializers.py:67]: So the `Y*Y` in `layernorm` requires float32, but after that we can continue doing everything in lower precision. Keeping `xn` in float32 inquires a slight perf loss, I think 3-4TFLOPS, if I remember correctly

Comment (chenyuxyz) [reply examples/mlperf/initializers.py:67]: `if not self.elementwise_affine: return xn.cast(x.dtype)` my question was why `cast(x.dtype)` only in this L67 and not the line below?

Comment (Eliulm) [reply examples/mlperf/initializers.py:67]: Oh sorry. Tbh, I missed that line. Will remove cast

**Comment by chenyuxyz** [comment]
tested on red (with DEBUG=0), looks good!

```
➜  tinygrad git:(4288) ✗ time BENCHMARK=10 GPUS=6 TRAIN_BEAM=4 EVAL_BEAM=4 WANDB=0 CKPT=1 CKPT_DIR=~/ckpts-final TRAIN_STEPS=200000 OPT_BASE_LEARNING_RATE=0.0001 MODEL=bert PYTHONPATH=. BASEDIR=/raid/datasets/wiki python3 examples/mlperf/model_train.py
training bert
Training on ['HSA:0', 'HSA:1', 'HSA:2', 'HSA:3', 'HSA:4', 'HSA:5']
HParam: "GPUS": ['HSA:0', 'HSA:1', 'HSA:2', 'HSA:3', 'HSA:4', 'HSA:5']
HParam: "seed": 12345
HParam: "GLOBAL_BATCH_SIZE": 24
HParam: "EVAL_BS": 24
HParam: "OPT_BASE_LEARNING_RATE": 0.0001
HParam: "TRAIN_STEPS": 200000
HParam: "NUM_WARMUP_STEPS": 20000
HParam: "MAX_EVAL_STEPS": 417
HParam: "EVAL_STEP_FREQ": 6250
HParam: "SAVE_CKPT_FREQ": 1000
HParam: "KEEP_CKPT_AMOUNT": 5
HParam: "INIT_CKPT_DIR": /raid/datasets/wiki
HParam: "decay": 0.01
HParam: "poly_power": 1.0
HParam: "DEFAULT_FLOAT": float
HParam: "TRAIN_BEAM": 4
HParam: "EVAL_BEAM": 4
Training with batch size 24 for 1 epoch with each 200000 steps
    0 103517.25 ms run, 103199.62 ms python,   4.00 ms fetch data,  313.63 ms HSA * 6,  6.28 loss, 0.000000 LR, 33.07 GB used,    282.49 GFLOPS
eval lm loss: 4.60, eval clsf loss: 1.90, eval lm accuracy: 0.344082,                   eval clsf accuracy: 0.53, avg eval step time: 0.17
    1 29347.79 ms run, 29298.17 ms python,  21.62 ms fetch data,   28.01 ms HSA * 6,  7.13 loss, 0.000000 LR, 67.64 GB used,    991.27 GFLOPS
    2 30582.29 ms run, 30544.95 ms python,   4.32 ms fetch data,   33.02 ms HSA * 6,  6.56 loss, 0.000000 LR, 109.34 GB used,    951.25 GFLOPS
    3  467.67 ms run,   81.20 ms python,   3.06 ms fetch data,  383.41 ms HSA * 6,  5.63 loss, 0.000000 LR, 109.34 GB used,  62205.40 GFLOPS
    4  460.48 ms run,   80.34 ms python,   4.20 ms fetch data,  375.93 ms HSA * 6,  6.20 loss, 0.000000 LR, 109.34 GB used,  63176.47 GFLOPS
    5  460.45 ms run,   80.92 ms python,   4.08 ms fetch data,  375.45 ms HSA * 6,  5.32 loss, 0.000000 LR, 109.34 GB used,  63180.77 GFLOPS
    6  460.09 ms run,   81.27 ms python,   4.68 ms fetch data,  374.14 ms HSA * 6,  5.86 loss, 0.000000 LR, 109.34 GB used,  63229.98 GFLOPS
    7  462.49 ms run,   80.90 ms python,   4.62 ms fetch data,  376.97 ms HSA * 6,  5.94 loss, 0.000000 LR, 109.34 GB used,  62901.32 GFLOPS
    8  459.15 ms run,   80.79 ms python,   4.58 ms fetch data,  373.78 ms HSA * 6,  6.64 loss, 0.000000 LR, 109.34 GB used,  63359.11 GFLOPS
    9  461.89 ms run,   81.58 ms python,   4.07 ms fetch data,  376.24 ms HSA * 6,  6.44 loss, 0.000000 LR, 109.34 GB used,  62983.65 GFLOPS
Estimated training time: 25h41m
epoch global_ops: 5_818_303_107_254_800_000, epoch global_mem: 101_987_445_707_200_000
```

**Comment by chenyuxyz** [comment]
oh that's before your last change (1 -> one). merging now
